File: year_2022/day_13/day13.py
from ast import literal_eval
from functools import cmp_to_key
from itertools import zip_longest

from tools import loader, parsers


def compare(left: list, right: list) -> int:
    if not isinstance(left, list): left = [left]
    if not isinstance(right, list): right = [right]

    for i, j in zip_longest(left, right):
        match i, j:
            case _, None:
                return 1
            case None, _:
                return -1
            case int(), int():
                if i != j:
                    return i - j
            case _:
                out = compare(i, j)
                if out != 0:
                    return out
    return 0


def part_1(data: list[list[str]]) -> int:
    """test part 1:
    >>> part_1(parsers.blocks('test13.txt'))
    13"""
    result = 0
    for index, pair in enumerate(data, start=1):
        if compare(*map(literal_eval, pair)) < 0:
            result += index
    return result


def part_2(data: list[str]) -> int:
    """test part 2:
    >>> part_2(parsers.lines('test13.txt'))
    140"""
    data.extend(('[[2]]', '[[6]]'))
    data = [literal_eval(line) for line in data if line]
    data.sort(key=cmp_to_key(compare))
    return (data.index([[2]]) + 1) * (data.index([[6]]) + 1)


print(part_1(parsers.blocks(loader.get())))  # 5252
print(part_2(parsers.lines(loader.get())))  # 20592

File: year_2022/day_13/day13.go
package main

import (
	"github.com/Nyaaa/advent-of-code/tools"
	"encoding/json"
	"fmt"
	"sort"
)

func unmarshal(str string) []any {
	result := []any{}
	_ = json.Unmarshal([]byte(str), &result)

	return result
}

func compare(left, right any) int {
	leftList, _ := left.([]any)
	rightList, _ := right.([]any)
	leftNum, leftIsNum := left.(float64)
	rightNum, rightIsNum := right.(float64)

	switch {
	case leftIsNum && rightIsNum:
		return int(leftNum - rightNum)
	case leftIsNum:
		leftList = []any{left}
	case rightIsNum:
		rightList = []any{right}
	}

	for i := 0; i < len(leftList) && i < len(rightList); i++ {
		if res := compare(leftList[i], rightList[i]); res != 0 {
			return res
		}
	}

	return len(leftList) - len(rightList)
}

func solve(input string) (int, int) {
	data := tools.SplitBlocks(tools.ReadLines(input))
	part1, part2 := 0, 1
	dataFull := []any{[]any{[]any{2.0}}, []any{[]any{6.0}}}

	for i, pair := range data {
		left := unmarshal(pair[0])
		right := unmarshal(pair[1])
		dataFull = append(dataFull, left, right)

		if compare(left, right) <= 0 {
			part1 += i + 1
		}
	}

	sort.Slice(dataFull, func(i, j int) bool {
		return compare(dataFull[i], dataFull[j]) <= 0
	})

	for i, line := range dataFull {
		if fmt.Sprint(line) == "[[2]]" || fmt.Sprint(line) == "[[6]]" {
			part2 *= i + 1
		}
	}

	return part1, part2
}

func main() {
	testData := "test13.txt"
	data := tools.GetData(2022, 13)

	fmt.Println(solve(testData))
	fmt.Println(solve(data))
}
